File: 0414. Third Maximum Number/0414.py
# time complexity: O(n)
# space complexity: O(1)
class Solution(object):
    # solution by using set
    def thirdMax(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        top_three = set()

        for num in nums:
            top_three.add(num) # add every number to the set
            if len(top_three) > 3: # if the length is greater than 3
                top_three.remove(min(top_three)) # remove the smallest element
        
        # if there are 3 distinct number, return the minimum
        # else return the maximum
        if len(top_three) == 3:
            return min(top_three)
        else:
            return max(top_three)

File: 0414. Third Maximum Number/0414.java
// time complexity: O(n)
// space complexity: O(1)
class Solution {
    public int thirdMax(int[] nums) {
        // Use long to handle edge cases where nums contains Integer.MIN_VALUE
        long firstMax = Long.MIN_VALUE;
        long secondMax = Long.MIN_VALUE;
        long thirdMax = Long.MIN_VALUE;

        for (int i = 0; i < nums.length; i++) {
            if (nums[i] > thirdMax && nums[i] < secondMax) {
                thirdMax = nums[i]; // update only thirdmax
            }
            else if (nums[i] > secondMax && nums[i] < firstMax) {
                thirdMax = secondMax; // update thirdmax and secondmax
                secondMax = nums[i];
            }
            else if (nums[i] > firstMax) {
                thirdMax = secondMax; // update thirdmax, secondmax and firstmax
                secondMax = firstMax;
                firstMax = nums[i];
            }
        }

        System.out.println(thirdMax);
        // If a third max exists, return it; otherwise, return the maximum number
        if (thirdMax != Long.MIN_VALUE) return (int)thirdMax;
        else return (int)firstMax;
    }
}
